### ISSUE TITLE
Unable to smart populate unowned 1:1 when autoJoinOneToOneOwner = false

### ISSUE THREAD

**Issue by emrosenf** [mikro-orm/mikro-orm#269]
I have `autoJoinOneToOneOwner: false` set in my config and am trying to smart populate an unowned 1:1 relationship. The relationship is not smart populated.

```typescript
const user = await ctx.em.findOneOrFail(User, { id, }, ['slug'])
```

Here is the relationship definition. `User` is not the owning side.

```typescript
class User {
...
  @OneToOne(() => Slug, slug => slug.user)
  slug: IdentifiedReference<Slug>;
}
```

Two deficiencies I notice in the current `autoJoinOneToOneOwner: false` behavior:

1. If I load a `User`, there is no key for `slug`. It's not even `null`, it's just not there. I expected it would be an `IdentifiedReference` with a `id = null`.

2. I expected to be able to override the default behavior by adding the relationship to the `populate` array. That does not seem to be possible at the moment.

**Comment by B4nan** [comment]
1. we can make it `null` value explicitly, but currently it is not designed the way you are expecting (with having `null` object always there). `Reference` instance always needs to have `id`. There is #264 that suggests the same, but I personally do not like it that much. Definitely not what I want to have as default behaviour, although I am thinking about supporting that optionally. 
2. you will still get instance of `Reference` wrapper, but yes, you should be definitely able to load the entity via `populate`. 

I guess I know what is happening, will try to look into it soon.

**Comment by B4nan** [comment]
Thinking about 1. more, and I think it is correct the key is not there. We do not know if it is null or not, it could be a value. Without the join, the information is just not present yet.

**Comment by emrosenf** [comment]
I think that’s a reasonable position, the main DX consideration was that it
made me second guess whether the relationship was properly defined. Being
able to populate it would have reassured me everything was ok

**Comment by B4nan** [comment]
I can't replicate this. Here is the test case: 

https://github.com/mikro-orm/mikro-orm/pull/271/files#diff-b5a1e90fb6b83d3756b42386aa6f9791R45

Do you see any difference from what you are doing?

**Comment by emrosenf** [comment]
I don't see any difference from what you are doing. I will take a break and look again.

But I don't see how your test case is possible. Please take a look at [EntityLoader.ts#158](https://github.com/mikro-orm/mikro-orm/blob/master/lib/entity/EntityLoader.ts#L158)

```typescript
  private getChildReferences<T extends AnyEntity<T>>(entities: T[], prop: EntityProperty<T>): AnyEntity[] {
    const filtered = this.filterCollections(entities, prop.name);
    const children: AnyEntity[] = [];

    if (prop.reference === ReferenceType.ONE_TO_MANY) {
      ...
    } else { // MANY_TO_ONE or ONE_TO_ONE
      children.push(...entities.filter(e => (Utils.isEntity(e[prop.name]) || e[prop.name] as unknown instanceof Reference) && !(e[prop.name] as AnyEntity).isInitialized()).map(e => {
        const ref = e[prop.name];
        return Utils.isReference<T>(ref) ? (ref as Reference<T>).unwrap() : ref;
      }));
    }
```

You have already agreed with me (on point 1) that `e[prop.name] === undefined`. So the `entities.filter` call will find nothing and at the end of the loop `children.length === 0`.

This will pop back up to `findChildren`

```typescript
  private async findChildren<T extends AnyEntity<T>>(entities: T[], prop: EntityProperty, where: FilterQuery<T> = {}, orderBy?: QueryOrderMap): Promise<AnyEntity[]> {
    const children = this.getChildReferences<T>(entities, prop);
    const meta = this.metadata.get(prop.type);
    let fk = meta.primaryKey;

    if (prop.reference === ReferenceType.ONE_TO_MANY || (prop.reference === ReferenceType.MANY_TO_MANY && !prop.owner)) {
      fk = meta.properties[prop.mappedBy].name;
    }

    if (children.length === 0) {
      return [];
    }
```

And return `[]`, which is used to initialize the collection

**Comment by B4nan** [comment]
That happens only if you do not provide the `populate` parameter explicitly. I thought that in your case it is broken because of the `Reference` wrapper, but it works (the new tests are passing, see #271. 

Also, the change hat introduced the `autoJoinOneToOneOwner`, was just about _not appending to the `populate` automatically_. In this case, the `populate` is already correctly formed (includes the 1:1 owner). 

But there is one possible case where I see it can be broken - if you had the `user` entity already loaded (without explicitly populating the owner). 

```typescript
const user123 = await ctx.em.findOneOrFail(User, 1); // can happen anywhere in the same context

// now user#1 is already in the identity map, so calling this will not load it from the database, but just try to populate the slug, which is in this case undefined
const user = await ctx.em.findOneOrFail(User, 1, ['slug']);
```

Maybe that is your problem? Can you doublecheck whether the user entity is already in the identity map before you do the `const user = await ctx.em.findOneOrFail(User, { id, }, ['slug'])` call? Just try to `console.log(ctx.em.getReference(User, 1))` and see if you get the full entity or just a reference with id.

**Comment by emrosenf** [comment]
You are 100% correct. This `User` is loaded during auth middleware, before my `findOneOrFail`. What do you suggest for this edge case?

**Comment by B4nan** [comment]
I will try to handle this somehow automatically. For findOne, you could pass { refresh: false } in the options param, that should trigger reloading of already known entity, but this time with the auto join happening.

**Comment by B4nan** [comment]
Fixed in `3.0.0-rc.4`.
